plugins/pur/core/src/main/java/org/pentaho/di/repository/pur/PurRepository.java

```diff
@@ -1165,7 +1264,12 @@ private ObjectId getObjectId( final String name, final RepositoryDirectoryInterf
     }
     List<RepositoryFile> allFiles = new ArrayList<RepositoryFile>();
     for ( String parentFolderPath : parentFolderPaths ) {
-      allFiles.addAll( pur.getDeletedFiles( parentFolderPath, mergedFilterBuf.toString() ) );
+      readWriteLock.readLock().lock();
```

Comment:
The cost of obtaining the Read Lock is actually pretty high compared to a simple mutex. So obtaining and releasing the lock should be done outside the loop, not inside.